tests/unit/s2n_build_test.c

```diff
@@ -0,0 +1,149 @@
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+#define _GNU_SOURCE
+#include <ctype.h>
+#include <openssl/crypto.h>
+#include <openssl/opensslv.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "s2n_test.h"
+
+struct libcrypto_info {
+    char *name;
+    int major_version;
+    int minor_version;
+    int patch_version;
+};
+
+int number_from_version(char *version)
+{
+    if (version == NULL) {
+        return -1;
+    }
+    if (version[0] == '0') {
+        return 0;
+    }
+    int version_number = atoi(version);
+    /* atoi returns 0 on error */
+    if (version_number == 0) {
+        return -1;
+    } else {
+        return version_number;
+    }
+}
+
+int convert_version_number_to_string(struct libcrypto_info *libcrypto_info, char *version_string_out)
+{
+    /* Format major minor and patch version in a normalized version string (e.x. "1.1.1", "3.0", "3")
+     * If there is no major version, return -1 and leave version_string_out untouched. */
+    if (libcrypto_info->major_version == -1) {
+        return -1;
+    }
+
+    if (libcrypto_info->major_version >= 0
+            && libcrypto_info->minor_version >= 0
+            && libcrypto_info->patch_version >= 0) {
+        sprintf(version_string_out, "%d.%d.%d", libcrypto_info->major_version, libcrypto_info->minor_version, libcrypto_info->patch_version);
+    } else if (libcrypto_info->major_version >= 0
+            && libcrypto_info->minor_version >= 0) {
+        sprintf(version_string_out, "%d.%d", libcrypto_info->major_version, libcrypto_info->minor_version);
+    } else if (libcrypto_info->major_version >= 0) {
+        sprintf(version_string_out, "%d", libcrypto_info->major_version);
+    }
+
+    return 0;
+}
+
+int extract_libcrypto_info_from_s2n_libcrypto(char *s2n_libcrypto_name, struct libcrypto_info *libcrypto_info)
+{
+    if (libcrypto_info == NULL || s2n_libcrypto_name == NULL || strlen(s2n_libcrypto_name) == 0) {
+        return -1;
+    }
+
+    libcrypto_info->name = NULL;
+    libcrypto_info->major_version = -1;
+    libcrypto_info->minor_version = -1;
+    libcrypto_info->patch_version = -1;
+
+    libcrypto_info->name = strtok(s2n_libcrypto_name, "-");
+    char *rest = strtok(NULL, "");
+
+    if (rest != NULL && isdigit(rest[0])) {
+        char *version_number_text = strtok(rest, ".");
+        libcrypto_info->major_version = number_from_version(version_number_text);
+        version_number_text = strtok(NULL, ".");
+        libcrypto_info->minor_version = number_from_version(version_number_text);
+        version_number_text = strtok(NULL, ".");
+        libcrypto_info->patch_version = number_from_version(version_number_text);
+    }
+
+    return 0;
+}
+
+int main()
+{
+    BEGIN_TEST();
+
+    const char *s2n_libcrypto = getenv("S2N_LIBCRYPTO");
+
+    /* S2N_LIBCRYPTO and S2N_BUILD_PRESET should be consistent.  */
```

Comment:
Should this test come after the `s2n_libcrypto == NULL` check?

Reply:
This first test says: If there is a s2n_build_preset it must match s2n_libcrypto. 

The `s2n_libcrypto == NULL` check conditionally ends the test if s2n_libcrypto isn't set.

Consider the following error condition: `s2n_libcrypto == NULL && s2n_build_preset != NULL`. If we end the test before the substring check we won't catch this condition.

Maybe I need a comment to clarify that?

Reply:
Ah okay, I didn't think of that case. That makes sense.